src/test/java/com/example/demo/api/ProductEndpointTest.java

```diff
@@ -1,97 +1,219 @@
+        //given
+        ProductRequestDto requestDto = new ProductRequestDto("testname", dummyPrice, dummyImg, new DescriptionDto(""), dummyTags);
+
+        //when
+        String requestJson = mapToJson(requestDto);
+        ResponseEntity<ProductResponseDto> response = httpClient.postForEntity(baseUrl(), getHttpRequest(requestJson), ProductResponseDto.class);
+
+        //then
+        assertThat(response.getStatusCodeValue()).isEqualTo(422);
     }
 
     @Test
     public void shouldGetExistingProduct() {
         //given
-        ProductRequestDto requestDto = new ProductRequestDto("name2");
+        ProductRequestDto requestDto = new ProductRequestDto("name2", dummyPrice, dummyImg, dummyDescr, dummyTags);
         ProductResponseDto existingProduct = productFacade.create(requestDto);
-        final String url = "http://localhost:" + port + "/products/" + existingProduct.getId();
+        String url = baseUrl() + "/" + existingProduct.getId();
 
         //when
         ResponseEntity<ProductResponseDto> result = httpClient.getForEntity(url, ProductResponseDto.class);
 
         //then
-        Assertions.assertThat(result.getStatusCodeValue()).isEqualTo(200);
-        Assertions.assertThat(result.getBody()).isEqualToComparingFieldByField(existingProduct);
+        assertThat(result.getStatusCodeValue()).isEqualTo(200);
+        assertThat(result.getBody()).isEqualToComparingFieldByField(existingProduct);
+    }
+
+    @Test
+    @DirtiesContext(methodMode = DirtiesContext.MethodMode.BEFORE_METHOD)
+    public void shouldGetExistingProductsByTags() {
+        //given
+        Arrays.asList(
+                new ProductRequestDto("name1", dummyPrice, dummyImg, dummyDescr, tagDtoSetFromTagNames("tag1", "tag4")),
+                new ProductRequestDto("name2", dummyPrice, dummyImg, dummyDescr, tagDtoSetFromTagNames("tag1", "tag2")),
+                new ProductRequestDto("name3", dummyPrice, dummyImg, dummyDescr, tagDtoSetFromTagNames("tag1", "tag3, tag4")),
+                new ProductRequestDto("name4", dummyPrice, dummyImg, dummyDescr, tagDtoSetFromTagNames("tag1", "tag4", "tag2"))
+        ).forEach(requestDto -> productFacade.create(requestDto));
+
+        String url = baseUrl() + "?tag=tag1&tag=tag2";
+        Set<String> expectedProductsNames = Set.of("name2", "name4");
+
+        //when
+        ResponseEntity<ProductListResponseDto> result = httpClient.getForEntity(url, ProductListResponseDto.class);
+        Set<String> fetchedProductsNames = result.getBody().getProducts().stream().map(ProductResponseDto::getName).collect(Collectors.toSet());
+
+        //then
+        assertThat(result.getStatusCodeValue()).isEqualTo(200);
+        assertThat(fetchedProductsNames.size()).isEqualTo(expectedProductsNames.size());
+        assertThat(fetchedProductsNames).containsAll(expectedProductsNames);
+    }
+
+    @Test
+    @DirtiesContext(methodMode = DirtiesContext.MethodMode.BEFORE_METHOD)
```

Comment:
Co to tutaj robi? :)

Reply:
Ten test nie przejdzie jeżeli wcześniejsze testy działały na bazie, dzięki tej adnotacji Spring postawi wszystko na nowo przed testem. Coś nie tak? ;)